### ISSUE TITLE
[macOS] Conflict handling of client ruins open SQLite files

### ISSUE THREAD

**Issue by daniel-hmc** [owncloud/client#5521]
Original issue: https://github.com/nextcloud/server/issues/3036
Closed there, because it's a client issue and those should be tracked here, I was told.

If an SQLite file, that's currently open in its application, is synced and the client resolves a conflict by downloading the server version of the file, saving it under the original file name in the local folder and renames the local file to "...conflict...", the application that uses the file has troubles accessing the file and data is lost.

### Expected behaviour
The client should not rename local files (because they could be in use). Instead, my suggestion would be that the server version of the file, that's downloaded, is saved under the local name "...conflict..." instead.

### Actual behaviour
explained above

### Steps to reproduce
1. Put as SQLite file (e.g. a file created with NoteCase Pro (www.notecasepro.com) using the ncdb file format (not the other formats, as they are not SQLite-based!) into an owncloud-synced directory. Leave the file open in NoteCase Pro. 
2. Trigger, that owncloud/nextcloud wants to sync the file in both directions (in my case, the reason was, that I moved the entire data repository to a new directory and updated the server; for some reason, occ wanted to sync "down" the server version of the file then. I have not modified the file on the server or any other client - a server bug? But not relevant here.)
3. Make changes to the local version of the file, to also trigger a sync
4. Start the owncloud client and let it sync the file, renaming the local version
5. In NoteCase Pro, file still open, make changes to the file and save (*)
6. See, that in NoteCase Pro the changes are not persistent (selecting another note and then the modified note again --> changes are gone!)

(*): Please do, like I did, a LOT of very important changes, about 2 hours of work. Then you get a feeling how frustrated I was when I finally realized that no data was saved! ;-)

### Server configuration
Operating system: Linux 4.7.0-0.bpo.1-amd64

Web server: 
Apache 2.4.10

Database:
MariaDB

PHP version:
5.6.29

ownCloud version:
Nextcloud 11.0.1

Storage backend (external storage):

### Client configuration
Client version:
Nextcloud client 2.2.4 (build 1)
 
Operating system:
Mac OS X 10.11.6

OS language:
de

Qt version used by client package (Linux only, see also Settings dialog):
--

Client package (From ownCloud or distro) (Linux only):
--

Installation path of client:
default:  /Applications/nextcloud.app

### Logs
n/a

**Comment by hodyroff** [comment]
"I moved the entire data repository to a new directory and updated the server" not necessarily a daily operation and something during which you should be in maintenance mode ... generally you are welcome to open issues here, but please reproduce them with ownCloud latest versions.

Handling of sqlite databases, interesting corner case. Good datapoint for now for the backlog.

**Comment by daniel-hmc** [comment]
Sure, that directory move is not a daily operation, and I was in maintenance mode. I mentioned this only in case it is relevant. The issue already happens, if you simply trigger a sync in both directions, i.e. stop clients on two synced machines and modify the same file on both machines, then start both clients again. And this IS a daily scenario.

So, if you don't consider this a high-priority issue (which I understand), do you have an idea how users can work around this issue for now? 

(BTW, does the "Feature" label mean you consider this a feature request? I'd rather consider this a bug report. In general I don't think it's a good idea for any sync client to rename a local file in the file system that could be in use by an application.
On the other hand, you have probably had good reasons to implement it that way and not in a different way.)

**Comment by hodyroff** [comment]
@daniel-hmc Yes. The design is indeed like this: 
https://doc.owncloud.org/desktop/2.2/architecture.html#comparison-and-conflict-cases
Generally if a sqlite database is manipulated at the different computers at the same time, what do you expect to happen? I can only think of a locking mechanism right now ... but you are welcome to suggest ideas.

**Comment by daniel-hmc** [comment]
Ah, I see: " The client resolves this conflict by creating a conflict file of the older of the two files and saving the newer file under the original file name." So not always the local file is renamed, but it is always the OLDER file that is renamed. Okay. But still, I think renaming a local file should be avoided.

# Suggestions:

S1: Introduce an (optional) **interactive conflict resolution** mode, in which conflicts are recorded during sync and at the end of the sync, a dialog is shown with a list of conflicts, in which the user can decide per conflict or for a group or all conflicts, how they are handled (options: use server file, use client file, use newer file, use oder file, use larger file, use smaller file, keep both files and change the server file's name to "...server...". Maybe even with an option to show the differences of file contents using diff....). most sync programs I know (and I know a lot) use such dialogs. However, most cloud sync clients, such as Dropbox, Wuala, Google Drive etc. don't. 

S2: **Always rename the *server* file to "...conflict..."**, not the *older* file (my initial suggestion).
This has the disadvantage, that in most cases the NEWER file will be the one the user wants to use in future, but the server file could also be the newer file, and then the user won't "automatically" use the correct file. i.e. the change that "by accident" the file with the original name is the file that the user wants to use is lower. But on the other hand, cases of data corruption are avoided, which is, IMHO, more important.

# Research:

**- Dropbox:** As I understand this https://www.dropbox.com/help/36, Dropbox has implemented conflict handling according to S2. (However, it's not entirely clear to me)

**- GDrive:** I haven't found any definite information about which of the conflicting files get the [conflict] tag in the file name.

**- OneDrive / SkyDrive:** http://windowsitpro.com/windows-8/skydrive-conflict-resolution "the file that was saved last will keep its original name, and the version that has been altered on the other machine will be renamed with the machine's name appended to the filename." --> also according to my S2, as it seems. 

**- iCloud:** https://developer.apple.com/library/content/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/iCloud/iCloud.html#//apple_ref/doc/uid/TP40010672-CH12-SW52 
" Its solution is to make the most recently modified file the current file and to mark any other versions of the file as conflict versions." --> same basic implementation as owncloud. BUT: Due to the more integrated way of applications interacting with iCloud in the Apple world, there are more recommended ways to resolve such issues: ..."Your app is notified of conflict versions through its file presenter objects. It is the job of the file presenter to decide how best to resolve any conflicts that arise. Apps are encouraged to resolve conflicts quietly whenever possible, either by merging the file contents or by discarding the older version if the older data is no longer relevant. However, if discarding or merging the file contents is impractical or might result in data loss, your app might need to prompt the user for help in choosing an appropriate course of action. For example, you might let the user choose which version of the file to keep, or you might offer to save the older version under a new name."

**Bottom line:**
It seems, that my S2 is not that uncommon, so why not implement it also in owncloud and nextcloud? There is this minor disadvantage of the slightly higher change that older data is accessible under the original file name and newer data is inside the "conflict" file. But IF there is a comflict, it's always better the user has a look at it anyway. Maybe the client should notify the user about any conflict file being generated with a button to go to the corresponding directory, so the user can have a look at the files and decide what to do with them.


-

**Comment by hodyroff** [comment]
@daniel-hmc Thanks for the analysis and research. We will gather further feedback over time and match it with customer and market expectations.
Generally dialogs and notifications or call to actions are tough because we are getting way more requests for headless operation right now. But will be reviewed as part of 3.0 conversations. Any further input welcome! Also put this on central.owncloud.org for people to chime in ...

**Comment by ckamm** [comment]
@daniel-hmc @hodyroff 
* Actually, the client always renames the local file to _conflict and puts the server's version in the expected location. This decision seems to have happened before May 2013; @dragotin would know. I'll update the documentation accordingly.
* Changing the conflict resolution strategy is possible, of course, but would have far-reaching consequences. Some existing users might very well depend on the current behavior.

**Comment by daniel-hmc** [comment]
@ckamm 
What about making a change optional (i.e. implement both ways and let the user choose via a setting)? Default behavior could remain as it is now, so users don't have to care about it. But once users run into problems like mine and WANT to switch, they can. The choice could be presented to the user when the first conflict is detected, with a short(!) message telling the user about the possible ways to resolve conflicts, their up- and downsides and a switch where the user can decide (with a default setting for "I am not sure", so users, who don't want to care feel safe).

**Comment by dragotin** [comment]
The decision to do it like this was made after long discussions. The idea is that the server version has precedence before the local version in any case. That is obviously not accurate in any situation, but whatever other solution you choose, there will be ways for similar frustration in a different situation. The problem with conflicts is that it means that there are two *parallel* changes to the file, and it is simply not possible to judge by an algorithm which change was more important.

**Comment by daniel-hmc** [comment]
Sure, but then a)  the user should get a chance to be notified about the situation and judge by himself (currently, conflicts are "resolved" silently, the user may only notice weeks later...) - that's the same with Dropbox & co, but why not make owncloud / nextcloud better? I'd feel much safer, if there was an option to bypass silent conflict resolution and instead being notified instantly, whenever a conflict is detected! I know, there are users who want it silent. That's why I suggest this to be an *option*.
And b) the local renaming of files has additional side effects / risks that could be avoided. So if it's not possible to judge automatically anyway, why not at least avoid additional risks?

**Comment by ogoffart** [comment]
So what should we do about this issue?
I don't think we should change the way we do the conflict resolution.

> The client should not rename local files (because they could be in use). 

Well, the files could be in use, and any change is potentially dangerous.
So the client should also not overwrite the file, nor remove it?  That wouldn't be so usefull as a sync client.

**Comment by ckamm** [comment]
I'd say we close this issue.

Unfortunately there's no reliable way for a sync client to detect whether a file is currently "in use" by an application. That means propagating remote changes will mean adjusting local files that may or may not be in use. Unless there's a specific reason that avoiding it for conflicts is essential I don't see why that would be worth it.

There could be a sync client that updates local files for incoming remote changes only on explicit user request but I doubt most users would appreciate that.

**Comment by hodyroff** [comment]
And now where we have WebDAV locking in place, it could actually be possible to lock the file prior to using it by the application via our API. Considering the alternatives we can close, yes. A database which is shared on two or more machines and changed on both sites at roughly the same time without a locking mechanism doesn't seem the best idea from an application perspective either I fear ...
